Sources/PointFree/HtmlCssInliner.swift

```diff
@@ -0,0 +1,157 @@
+}
+
+/// Transforms `node` by applying a reasonable number of the styles from `stylesheet` to each element.
+public func applyInlineStyles(node: Node, stylesheet: Stylesheet) -> Node {
+  switch node {
+  case .comment:
+    return node
+
+  case let .document(nodes):
+    return .document(applyInlineStyles(nodes: nodes, stylesheet: stylesheet))
+
+  case let .element(element):
+    return .element(applyInlineStyles(element: element, stylesheet: stylesheet))
+
+  case .text:
+    return node
+  }
+}
+
+private func applyInlineStyles(element: Element, stylesheet: Stylesheet) -> Element {
+  let currentStyles = element.attribs
+    .first(where: { $0.key == "style" })?
+    .value
+    .renderedValue()?
+    .string
+    ?? ""
+
+  // Computes all inline styles based on the classes on this element
+  let classStyles = (
+    element.attribs
+      .first(where: { $0.key == "class" })?
+      .value
+      .renderedValue()?
+      .string
+      ?? ""
+    )
+    .components(separatedBy: " ")
+    .map { inlineStyles(for: .class($0), stylesheet: stylesheet) }
+    .joined(separator: ";")
+
+  // Computes all inline styles based on the name of the element tag.
+  let elemStyles = inlineStyles(for: .elem(.other(element.name)), stylesheet: stylesheet)
+
+  // Computes all inline styles based on the id of the element.
+  let idStyles = (
+    element.attribs
+      .first(where: { $0.key == "id" })?
+      .value
+      .renderedValue()?
+      .string
+    )
+    .map { inlineStyles(for: .id($0), stylesheet: stylesheet) }
+    ?? ""
+
+  let newStyles = [
+    currentStyles,
+    classStyles,
+    elemStyles,
+    idStyles
+    ]
+    .filter { !$0.isEmpty }
+    .joined(separator: ";")
+
+  let newAttribs = element.attribs
+    .filter { $0.key != "style" }
+    + (newStyles.isEmpty ? [] : [AnyAttribute("style", newStyles)])
+
+  return element
+    |> \.attribs .~ newAttribs
+    |> \.content .~ applyInlineStyles(nodes: element.content ?? [], stylesheet: stylesheet)
+}
+
+// Computes the inline styles for a selector given a stylesheet.
+private func inlineStyles(for selector: CssSelector, stylesheet: Stylesheet) -> String {
+  return stylesheet.rules // TODO: make rules public
+    .lazy
+    .map { inlineStyles(for: selector, rule: $0) }
+    .filter { !$0.isEmpty }
+    .joined(separator: ";")
+}
+
+// Computes the inline styles for a selector given a single styling rule.
+private func inlineStyles(for selector: CssSelector, rule: Rule) -> String {
+  switch rule {
+  case let .property(key, value):
+    return Css.rule(inline, [], [(key, value)])
+
+  case let .nested(app, rules):
+    if doesApp(app, contain: selector) {
+      return rules
+        .map { inlineStyles(for: selector, rule: $0) }
+        .joined(separator: ";")
+    }
+    return ""
+
+  case .query, .face, .keyframe, .import:
+    // TODO: what to do about `query` here? Media queries can't be used in inline styles, but we might still
+    //       want to render its styles?
```

Comment:
Huh! This comment makes me wonder where inlining styles in attributes is better than just embedding `style` tags.

Reply:
some email clients support style tags, but sometimes only a subset of things inside that tag. and even with the style tag things can get overridden by the client if the style isnt actually inlined. it's a wild world! 

thinking we are in a good sitch by just rendering inline and providing the style tag. espesh since we use 0 cascading in our stylesheets